### ISSUE TITLE
xonsh memory usage - suggestion threaded/client server model

### ISSUE THREAD

**Issue by eskhool** [xonsh/xonsh#2410]
If one has multiple instances of xonsh running like it happens with shells...each forked process takes around 45 Mb of resident memory. I am guessing this is due to the python3 process that each runs. I know the argument that memory/cpu is cheap nowadays but the shell especially as a login shell needs to be performant. a shell like fish takes ~5MB. 10 shells running takes 1/2 GB of RAM!

One solution could be using a client server model where a very light weight frontend just handling the terminal input (maybe in C) attaches via IPC to an existing running xonsh process which spawns a thread for it much like emacs. This would be lighter both for startup times and total memory consumption

I understand that this does introduce a risk of a catastrophic crash into the server xonsh process bringing down running shell but it merit a discussion and hence this issue.

**Comment by TeamSpen210** [comment]
One problem with that is that xonsh is also a Python interpreter, and each shell needs to keep that seperate. There isn't much you can share without causing problems.

**Comment by eskhool** [comment]
Sure xonsh is a python interpreter but couldn't it service multiple threads, each communicating with its own tiny shell frontend executable? This doesn't have to be the default but definitely needed for people with limited ram machine since each xonsh instance is taking an inordinate amount of ram and definitely causing a slowdown at startup also.

**Comment by astronouth7303** [comment]
Global state gets in everywhere. Environment variables, current directory, any xontribs using global variables, xonsh configuration, loaded modules, etc.

This would require a massive rewrite and re-architecture, basically changing how everything works forever. Nothing would be safe.

We'd have to work really hard to keep different shell contexts isolated, and having things like multiple environments would break this entirely. (There's been discussion about if we could "load" a venv environment into the shell context and/or "enter" it with a new shell.) Already, the shell context is 0% isolated from the guts of xonsh, and we already have problems with how easy it is to completely break it.

That is not to say there isn't room for improvement in these areas, but maybe we should try some of the more nuanced and informed approaches first.

I'm closing this as "ain't gonna happen". While it might _strictly_ help, it's got huge problems and we don't have the man power or commitment to make sweeping changes like this.

@eskhool If you'd like help reducing xonsh's footprint on your system, please feel free to reach out to us on [gitter](https://gitter.im/xonsh/xonsh) and [irc](irc://irc.oftc.net/#xonsh).

**Comment by scopatz** [comment]
Thanks for bringing this up @eskhool!  Just to add my two cents here, I think that moving to an interprocess model for history is an intriguing and good idea.  This has come up before with respect to sharing commands between sessions. In theory, I am all for it!  However, I think this feature needs a champion. @eskhool - if you'd be willing to take on that role, that would be awesome!

On the other end, you seems to be talking about thin front end clients for all of xonsh to a single interpreter. I am not a huge fan of this for a couple of reasons:

1. We sort of already have this if you use the Xonsh Jupyter kernel, no need to reinvent the wheel.
2. Each xonsh session being its own process makes management simple and tasks like killing a session easy and standard.

We might be able to get some startup time increases, etc. out of moving to this kind of model. But in my opinion it isn't worth the effort at this time. 

tl;dr Go for it for history!

**Comment by eskhool** [comment]
Hi @scopatz, I do get that the core dev community by and large is not feeling the performance difference (in order: memory usage, startup time etc with xonsh). I use a relatively recent Macbook pro retina with 8Gb RAM and find the lag on launching a new iterm session irritating since I spawn and close those very often, I can only assume that they are ok with the trade off. For my servers, the memory consumption is quite significant for each xonsh session and if the history sharing which is a significant effort anyways will still leave a 15Mb app minimum (guessing based on the size of the python interpreter and basic xonsh code)

The way I see it, for 90% of everyday muscle memory tasks that a shell involves launching a large dynamic interpreter like python (which i love and use everyday ) is overkill.
However when I do need some logic I don't like to pick up (esoteric bash/(some other new fangled shell language however good it may be in principle - i am looking at you fish). Then I want the entire python ecosystem at my fingertips. So while the idea of xonsh is great, the performance is not with or without history sharing will be not.

So the real question is IF I am able to work on a branch with the following basic design decision, is the opposition to the idea sufficient that it would get automatically ruled out for inclusion in the core ? or is it only a man hours issue?

I ask this because it changes the way the code needs to be approached. For inclusion, I will try to keep it similar to core and regularly backport merge the mainline trunk. Otherwise, it would be easier to take a clean break and not have to keep the codebases sympatico.

**Comment by eskhool** [comment]
Also @scopatz, as far as process/task management is concerned, the task is to make xonsh multi-thread safe. Beyond that the same concerns will apply as before, with a few easier ones of IPC etc. 

I understand writing a shell and supporting it is no mean task (kudos to you for finally doing it with python) but if it is to work for thousands of devops, all those extra seconds and hundreds of MBs of RAM saved will definitely add up to make it worth it.
